registration/api/v2/serializers.py

```diff
@@ -315,6 +344,88 @@ def validate_user(self, value):
+        control of a client resource (domain) if they forget to restrict its access.
+        """
+        domain_data = tldextract.extract(value)
+        domain = domain_data.registered_domain
+
+        if domain == settings.DEFAULT_INSTANCE_BASE_DOMAIN:
+            raise ValidationError('The domain "{domain}" is not allowed.'.format(
+                domain=value
+            ))
+
+        self._validate_subdomain_not_contains_reserved_word(domain_data.subdomain)
+
+        is_domain_used_for_beta_app = BetaTestApplication.objects.filter(
+            external_domain__endswith=domain
+        ).exists()
+
+        if is_domain_used_for_beta_app:
+            raise ValidationError('This domain is already taken.')
+
+        is_taken = OpenEdXInstance.objects.filter(
+            Q(external_lms_domain=domain)
+            | Q(external_lms_preview_domain__endswith=domain)
+            | Q(external_studio_domain__endswith=domain)
+            | Q(external_discovery_domain__endswith=domain)
+            | Q(external_ecommerce_domain__endswith=domain)
+            # No need to check for subdomain, since it will match anyway
+            | Q(extra_custom_domains__contains=domain)
+        ).exists()
+
+        if is_taken:
+            raise ValidationError('This domain is already taken.')
+
+        return value
+
+    def validate_subdomain(self, value) -> str:
+        """
+        Prevent users from registering with a subdomain which is in use.
+
+        The validation reduces the risk of security issues when someone is trying to take over
+        control of a client resource (domain) if they forget to restrict its access.
+        """
+        truncated_subdomain = value.split(".")[0]
+        self._validate_subdomain_not_contains_reserved_word(truncated_subdomain)
+
+        # Although the validator on the model would catch the existence of
+        # the beta application with the given subdomain, this way we can potentially
+        # save an API request to Gandi for checking the existance of the DNS record.
+        is_subdomain_used_for_beta_app = BetaTestApplication.objects.filter(
+            subdomain=value
+        ).exists()
+
+        if is_subdomain_used_for_beta_app:
+            raise ValidationError("This domain is already taken.")
+
+        domains = set([
+            settings.DEFAULT_INSTANCE_BASE_DOMAIN,
+            settings.GANDI_DEFAULT_BASE_DOMAIN,
+        ])
+
+        for domain in domains:
+            try:
+                managed_domains = gandi_api.filter_dns_records(domain)
+                records = {tldextract.extract(record["content"]) for record in managed_domains}
+            except Exception as exc:
+                logger.warning("Unable to retrieve the domains for %s: %s.", domain, str(exc))
+                raise ValidationError("The domain cannot be validated.")
+
+            # Because manually registered CNAMEs may have a dot (.) in their subdomain
```

Comment:
note that subdomain nesting is infinite... nothing is stopping someone registering `a.b.c.d.e.f.g.h.i.example.com`.

Reply:
Hm, the wording here maybe not the best, though there is an existing test case (as far as I remember) and an attached sample for that ((3) Subdomain of an existing instance's domain <- in the description) so that's checked as well and working. I'll rephrase the comment

Reply:
This is what you thought about, right?

Reply:
Rephrased